### PR TITLE
add missing link to libclang-cpp, sort link list in 'bin/'

### PR DESCRIPTION
**Pull request by sknigh**

### CODE REVIEW

File: bin/Makefile.am [hunk 1]
```diff
@@ -114,24 +114,23 @@ endif
...
  -Wl,-rpath,@CLANG_INSTALL_DIR@/lib \
- -lclangFrontend \
- -lclangFrontendTool \
- -lclangTooling \
- -lclangBasic \
+ -lclangAnalysis \
+ -lclangAST \
  -lclangASTMatchers \
+ -lclangBasic \
+ -lclang-cpp \
```
Comment (jjwilke) [inline bin/Makefile.am]: We will need to add logic that only adds this >= 9.

**Comment by jjwilke** [comment]
@sknigh Please retest with latest changes

File: acinclude/check_clang.m4 [hunk 2]
```diff
@@ -69,6 +69,7 @@ AC_DEFUN([CHECK_CLANG_LLVM], [
 
   if test "X$found_clang" = "Xno"; then
     AM_CONDITIONAL(HAVE_CLANG, false)
+    AM_CONDITIONAL(CLANG_NEED_LIPCPP,false)
```
Comment (jpkenny) [inline acinclude/check_clang.m4]: LIPCPP?

File: acinclude/check_clang.m4 [hunk 3]
```diff
@@ -86,6 +87,14 @@ AC_DEFUN([CHECK_CLANG_LLVM], [
     LLVM_CPPFLAGS="$CLANG_CPPFLAGS"
     LLVM_LDFLAGS="$CLANG_LDFLAGS"
 
+    clang_version=`$clang/bin/clang --version | head -n 1 | cut -d ' ' -f 3`
+    clang_major_version=`echo $clang_version | cut -d '.' -f 1`
+    if test "$clang_major_version" = "9"; then
+      AM_CONDITIONAL(CLANG_NEED_LIPCPP,true)
```
Comment (jpkenny) [inline acinclude/check_clang.m4]: Please change LIPCPP to LIBCPP

File: acinclude/check_clang.m4 (hunk 2 above)
Comment (jpkenny) [inline acinclude/check_clang.m4]: Please change LIPCPP to LIBCPP

File: acinclude/check_clang.m4 [hunk 4]
```diff
@@ -86,6 +87,14 @@ AC_DEFUN([CHECK_CLANG_LLVM], [
...
+    else
+      AM_CONDITIONAL(CLANG_NEED_LIPCPP,false)
```
Comment (jpkenny) [inline acinclude/check_clang.m4]: Please change LIPCPP to LIBCPP

File: bin/Makefile.am [hunk 5]
```diff
@@ -114,26 +114,29 @@ endif
...
+ -lclangTooling \
  -lclangToolingCore \
- -lclangAST \
- -lclangRewrite \
- -lclangLex \
- -lclangBasic \
  @CLANG_LIBTOOLING_LIBS@ \
  @CLANG_LIBTOOLING_SYSTEM_LIBS@
+
+if CLANG_NEED_LIPCPP
```
Comment (jpkenny) [inline bin/Makefile.am]: Please change LIPCPP to LIBCPP

**Comment by sknigh** [comment]
@jjwilke This didn't solve the issue

**Comment by jpkenny** [comment]
> @jjwilke This didn't solve the issue

@sknigh can you elaborate?

**Comment by sknigh** [comment]
Still testing... I am getting the linker errors with with LLVM 9 after your commit.

**Comment by jpkenny** [comment]
@sknigh not impossible that there is an ordering dependency, I don't fully understand why, but some linkers are more problematic than others -- try putting libcpp last and see if that helps.

**Comment by sknigh** [comment]
It's a problem with check_clang.m4. It fails when I use `--with-clang` without a path, but works when I provide one. When a path is not provided, the configure script also prints this on stderr, but finishes with clang enabled anyway
```
../configure: line 23547: yes/bin/llvm-config: No such file or directory
../configure: line 23548: yes/bin/llvm-config: No such file or directory
../configure: line 23554: yes/bin/clang: No such file or directory
Traceback (most recent call last):
  File "../bin/config_tools/get_clang_includes", line 18, in <module>
    lines = regexp.search(vOutput).groups()[0].strip().splitlines()
AttributeError: 'NoneType' object has no attribute 'groups'
Traceback (most recent call last):
  File "../bin/config_tools/get_clang_includes", line 18, in <module>
    lines = regexp.search(vOutput).groups()[0].strip().splitlines()
AttributeError: 'NoneType' object has no attribute 'groups'
```

**Comment by sknigh** [comment]
#347 related?

**Comment by jpkenny** [comment]
@sknigh, yes, I'd say related

**Comment by jjwilke** [comment]
Canceled in favor of PR #508

**Comment by calewis** [comment]
@jjwilke did you actually cancel this?

**Comment by jjwilke** [comment]
Yes. Updated PR is #508

**Pull request closed without merging**